server/src/main/java/org/apache/cassandra/sidecar/job/Job.java

```diff
@@ -0,0 +1,152 @@
+package org.apache.cassandra.sidecar.job;
+
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.sidecar.common.utils.JobResult;
+import org.apache.cassandra.sidecar.common.utils.JobResult.JobStatus;
+
+/**
+ * An abstract class representing a Job managed by the sidecar.
+ *
+ */
+public abstract class Job
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(Job.class);
+
+    protected UUID jobId;
+    protected JobStatus status;
+    protected String failureReason;
+    final CountDownLatch latch = new CountDownLatch(1);
+
+    protected Job()
+    {
+
+    }
+    /**
+     * Constructs a job with a unique UUID, in Pending state
+     * @param jobId UUID representing the Job to be created
+     */
+    protected Job(UUID jobId)
+    {
+        this.jobId = jobId;
+        this.status = JobStatus.Pending;
+        this.failureReason = "";
+    }
+
+    @VisibleForTesting
+    protected Job(UUID jobId, JobStatus status)
+    {
+        this.jobId = jobId;
+        this.status = status;
+        this.failureReason = "";
+    }
+
+    public void setStatus(JobStatus status)
+    {
+        this.status = status;
+    }
+    public JobStatus status()
+    {
+        return status;
+    }
+    public String failureReason()
+    {
+        return failureReason;
+    }
+
+    public UUID jobId()
+    {
+        return jobId;
+    }
+
+    /**
+     * Supplier specifying the functionality of the job to be triggered when the job is executed. Subclasses to
+     * provide operation-specific implementations
+     * @return a function with the operation implementation that returns a {@code JobResult}
+     * @throws Exception
+     */
+    public abstract Supplier<JobResult> jobOperationSupplier() throws Exception;
+
+    /**
+     * Provide a meaningful name of the operation executed by the concrete subclass.
+     * @return the name of the operation. eg. nodetool command name
+     */
+    public abstract String operation();
+
+    /**
+     * Specifies the downstream operation to be performed to check if the job is running on the Cassandra node/cluster.
+     * This functionality is provided by the Job when it is asynchronously triggering the job via the {@code JobManager}
+     * For synchronous jobs this should always return false.
+     * @return true if the job is running downstream
+     */
+    public abstract boolean checkInflightJob();
+
+    /**
+     * Execute the job behavior as specified in the operation supplier {@link #jobOperationSupplier()},
+     * while tracking the status of the job's lifecycle.
+     */
+    public void execute()
+    {
+        try
+        {
+            LOGGER.info("Executing job with ID: {}", jobId);
+            JobResult result = jobOperationSupplier().get();
```

Comment:
it feels we should use Future primitives here instead, execute should return  a future as well, instead of doing latch countdown

Reply:
Addressed in latest revision